### ISSUE TITLE
Usage of custom expressions in drift files

### ISSUE THREAD

**Issue by laurensfischer** [simolus3/moor#3366]
Hey, still a big fan of your library!

we are currently working on migrating an app to the newest drift version, coming from 1.x.

This works fine after upgrading for most queries, but we get an error when using a custom expression within the SELECT statement. The error while generating is `Null check operator used on a null value`.

We are mainly doing this to reuse the expression (and some code) in multiple queries and drift files.

Is it not possible anymore to use custom expression in SELECT statements?

If not, are there better alternatives than calling the custom functions directly in drift files? Doing that shows a warning (` Function calcStatus could not be found`), but the generated code looks okay. 

### Example
Simplified example of our queries and how it worked before upgrading and the full error that occurs now:

Query:
``` 
exampleQuery as Example:
    SELECT CAST($status AS INT) AS status FROM examples;
```

Generated code:
```
  Selectable<Example> exampleQuery(
      Expression status) {
    final generatedstatus = $write(status, hasMultipleTables: true);
    return customSelect(
        'SELECT CAST(${generatedstatus.sql} AS INT) AS status',
        variables: [
          ...generatedstatus.introducedVariables,
        ],
        readsFrom: {
          examples,
          ...generatedstatus.watchedTables,
        }).map((QueryRow row) {
      return Example
        status: row.read<int>('status'),
      );
    });
  }
}
```

Caller:
```
  Selectable<Example> getExample(
    int param,
  ) {
    return exampleQuery(CustomExpression(
      "calcStatus($param)",
    ));
  }
```

calcStatus is a custom function added to sql by calling `database.createFunction(functionName: "calcStatus", ...)`

The full error:

Error:
```
Null check operator used on a null value
package:drift_dev/src/writer/queries/utils.dart 29:60           FoundElementType.dartType.<fn>
package:drift_dev/src/analysis/results/dart.dart 45:10          new AnnotatedDartCode.build
package:drift_dev/src/writer/queries/utils.dart 10:32           FoundElementType.dartType
package:drift_dev/src/writer/queries/query_writer.dart 400:40   QueryWriter._writeParameters.typeFor
package:drift_dev/src/writer/queries/query_writer.dart 432:20   QueryWriter._writeParameters
package:drift_dev/src/writer/queries/query_writer.dart 300:5    QueryWriter._writeSelectStatementCreator
package:drift_dev/src/writer/queries/query_writer.dart 82:5     QueryWriter._writeSelect
package:drift_dev/src/writer/queries/query_writer.dart 71:7     QueryWriter.write
package:drift_dev/src/writer/drift_accessor_writer.dart 45:39   AccessorWriter.write
package:drift_dev/src/backends/build/drift_builder.dart 446:49  _DriftBuildRun._generateMonolithic
package:drift_dev/src/backends/build/drift_builder.dart 184:7   _DriftBuildRun.run
package:drift_dev/src/backends/build/drift_builder.dart 109:5   DriftBuilder.build
```

**Comment by simolus3** [comment]
This is still supposed to work :) The problem is the cast expression, which means that we don't know what type `$status` is supposed to have. So we can't generate an `Expression<int>`-typed parameter. But instead of crashing, we should just generate `Expression` without type parameters instead.

**Comment by simolus3** [comment]
I've fixed the problem in 53b5afcda7005565b3c46f0df65fea44730d0ec0. You could [try using that version](https://drift.simonbinder.eu/Internals/) of `drift_dev` to see if you run into any other issues after upgrading, I'll release the fix to pub.dev later today.

**Comment by simolus3** [comment]
`drift_dev` version `2.22.1` has just been published to pub.dev and contains this fix.
